bredis/AsyncConnection.hpp

```diff
@@ -25,9 +25,11 @@
 
 namespace bredis {
 
+// Rename `S` to `NextLayer` or `AsyncStream` 
 template <typename S> class AsyncConnection {
     using protocol_type_t = typename S::protocol_type;
 
+    // What is the reason for constraining the type of S this way?
     static_assert(std::is_same<protocol_type_t, boost::asio::ip::tcp>::value ||
                       std::is_same<protocol_type_t,
                                    boost::asio::local::stream_protocol>::value,
```

Comment:
That is true, but constraining the choice of the next layer type prevents interesting use-cases. For example, lets say that someone wants to measure the bytes transferred in and out on the bredis connection. They don't need to modify your code, they could instead create an **SyncStream** or **AsyncStream** wrapper around a real TCP/IP socket which counts bytes in calls to `read_some` and `write_some`. Your static assertion would prevent this use, even though it doesn't break invariants. A more sophisticated version of this proposed wrapper could implement bandwidth control. Your users could extend the functionality of your connection objects without changing the source code.

This is beyond theoretical, Beast already comes with a very useful `fail_stream` wrapper. With it, you could declare `bredis::AsyncConnection<beast::test::fail_stream<boost::asio::ip::tcp::socket>>` in a test, and exercise all the possible failure modes to increase test coverage.